scipy/linalg/tests/test_basic.py

```diff
@@ -931,35 +930,75 @@ def test_check_finite(self):
 
 class TestDet:
     def setup_method(self):
-        np.random.seed(1234)
+        self.rng = np.random.default_rng(1680305949878959)
 
     def test_simple(self):
         a = [[1, 2], [3, 4]]
-        a_det = det(a)
-        assert_almost_equal(a_det, -2.0)
+        assert_allclose(det(a), -2.0)
+        a = np.flipud(a)
+        assert_allclose(det(a), 2.0)
 
     def test_simple_complex(self):
         a = [[1, 2], [3, 4j]]
-        a_det = det(a)
-        assert_almost_equal(a_det, -6+4j)
+        assert_almost_equal(det(a), -6+4j)
 
     def test_random(self):
-        basic_det = linalg.det
         n = 20
-        for i in range(4):
-            a = random([n, n])
-            d1 = det(a)
-            d2 = basic_det(a)
-            assert_almost_equal(d1, d2)
+        a = self.rng.random([n, n])
+        d1, d2 = det(a), np.linalg.det(a)
+        assert_allclose(d1, d2)
+
+    def test_random_stacked(self):
+        n = 20
+        a = self.rng.random([3, 2, n, n])
+        d1, d2 = det(a), np.linalg.det(a)
+        assert_almost_equal(d1, d2)
 
     def test_random_complex(self):
-        basic_det = linalg.det
         n = 20
-        for i in range(4):
-            a = random([n, n]) + 2j*random([n, n])
-            d1 = det(a)
-            d2 = basic_det(a)
-            assert_allclose(d1, d2, rtol=1e-13)
+        a = self.rng.integers(-2, 3, size=[n, n])*0.25j
+        a += self.rng.random([n, n])
+        d1, d2 = det(a), np.linalg.det(a)
+        assert_almost_equal(d1, d2)
+
+    def test_random_complex_stacked(self):
+        n = 20
+        a = self.rng.integers(-2, 3, size=[3, 2, n, n])*0.25j
+        a += self.rng.random([3, 2, n, n])
+        d1, d2 = det(a), np.linalg.det(a)
+        assert_almost_equal(d1, d2)
+
+    # g and G dtypes are handled differently in windows and other platforms
+    @pytest.mark.parametrize('typ', [x for x in np.typecodes['All'][:20]
+                                     if x not in 'gG'])
+    def test_sample_compatible_dtype_input(self, typ):
+        n = 4
+        a = self.rng.random([n, n]).astype(typ)  # value is not important
+        assert isinstance(det(a), (float, complex))
```

Comment:
This is basically to prevent blowups in the single precision, like the example above, 

```
np.linalg.det(np.diag([1.e38, 1e38]).astype(np.float32))
```

This is a finite number that fits into double precision but gives inf in the single precision. So regardless of the input we upcast to double real or complex precision. I wasn't sure about this but it is the existing behavior unfortunately.

Reply:
There are a few other questions above. I've added numbers so they're easier to see. It looks like you've replied to no. 3 only.

Reply:
Yes, because I'm thinking about the first two :) 

1 - One of the strange behaviors is the silent up/down casting. So probably we should either let it blowup or mention that it would always be up/downcasted. 
2- You mean I should strip the array from the scalars? I should check what the cases are quickly

Reply:
2. is now fixed to np.complex128 and np.float64